janus_core/src/test_util/kubernetes.rs

```diff
@@ -2,21 +2,92 @@
 
 use kube::config::{KubeConfigOptions, Kubeconfig};
 use rand::{thread_rng, Rng};
-use std::process::{Command, Stdio};
-use tempfile::{NamedTempFile, TempPath};
+use std::{
+    path::{Path, PathBuf},
+    process::{Child, Command, Stdio},
+};
+use tempfile::NamedTempFile;
+use tracing::debug;
+
+/// Cluster represents a running Kubernetes cluster.
+pub struct Cluster {
+    /// Path to the Kubernetes config file, e.g., `~/.kube/config`
+    kubeconfig_path: PathBuf,
+    /// Name of the `kubectl` context to use.
+    context_name: String,
+}
+
+impl Cluster {
+    /// Creates a handle to an existing Kubernetes cluster.
+    pub fn create<P: AsRef<Path>>(kubeconfig_path: P, context_name: &str) -> Self {
+        let kubeconfig_path = kubeconfig_path.as_ref();
+        debug!(?kubeconfig_path, context_name, "Creating cluster handle");
+        Self {
+            kubeconfig_path: kubeconfig_path.to_path_buf(),
+            context_name: context_name.to_owned(),
+        }
+    }
+
+    /// Returns a new [`kube::Client`] configured to interact with this Kubernetes cluster.
+    pub async fn client(&self) -> kube::Client {
+        kube::Client::try_from(
+            kube::Config::from_custom_kubeconfig(
+                Kubeconfig::read_from(&self.kubeconfig_path).unwrap(),
+                &KubeConfigOptions {
+                    context: Some(self.context_name.clone()),
+                    ..KubeConfigOptions::default()
+                },
+            )
+            .await
+            .unwrap(),
+        )
+        .unwrap()
+    }
+
+    /// Set up port forwarding from `local_port` to `service_port` on the service in the namespace.
+    /// Returns a [`PortForwardGuard`] which will kill the child `kubectl` process on drop.
+    pub fn forward_port(
+        &self,
+        namespace: &str,
+        service: &str,
+        local_port: u16,
+        service_port: u16,
+    ) -> PortForwardGuard {
+        let args = [
+            "--kubeconfig",
+            &self.kubeconfig_path.to_string_lossy(),
+            "--context",
+            &self.context_name,
+            "--namespace",
+            namespace,
+            "port-forward",
+            &format!("service/{service}"),
+            &format!("{local_port}:{service_port}"),
+        ];
+        debug!(?args, "Invoking kubectl");
+        let child = Command::new("kubectl")
```

Comment:
Using this is inherently racy: the child process takes some time to set up port forwarding.

Maybe tail stdout for a while until you see the expected logging, or alternatively poke the port for a while until something is accepting connections on the port? (the former depends on `kubectl`'s output being stable, I believe the latter assumes that something will already be listening behind the forwarded port; maybe a cleverer solution is possible?)

Reply:
My plan for dealing with this race is to resolve https://github.com/divviup/janus/issues/196. IMO polling on a port or watching for logs doesn't reliably solve the problem--the process might still not be ready to handle messages. We could block until the Kubernetes health or liveness probe on the pod or service succeeds, but that doesn't actually help with the race with `kubectl port-forward`. I think the only way to do this reliably is for the HTTP clients to do reasonable retries if they encounter connection refused errors. Hence I argue for resolving this by fixing #196 instead of doing something in the test harness.

Reply:
I think looking to the logging does reliably solve the problem of detecting when the port forward is "ready" as far as `kubectl port-forward` is concerned, since it will only log once the port is "ready". Attempting to connect to the port would work to detect readiness too, as long as (a) there is some process in K8S to listening on the forwarded port (false negative) and (b) nothing else is listening on that port yet (false positive).

If the underlying process running in K8S accepts connections but isn't ready to handle them, that's a bug in the underlying process -- we should just fix the bug in the process, which we can do since we are writing the software under test. (That is, a correct server won't start listening to accept requests until it can handle them.) More to the point, I'd like to not have to add retries to all of our test-setup requests.

Reply:
What I'm worried about is that no matter what kind of preflight checks we do on the availability and functionality of the `port-forward`, the component that ends up making HTTP requests to the forwarded port must be robust against transient failures. Otherwise you end up with TOCTOU races. So if we have to deal with that robustness problem anyway, why not spare ourselves the hassle of writing the preflight checks (which would be brittle anyway -- the string that `kubectl` logs is not an API contract)?

Reply:
> the component that ends up making HTTP requests to the forwarded port must be robust against transient failures

My argument is that:
(a) this is true for non-test request code running in production (though even then a direct retry may or may not be the right answer, depending on the request)
(b) ...but it's definitely not true for test-only code (such as code adding the task under test & doing other test setup), which are generally going to be talking over IPC (for in-process or Kind-based tests), which we can assume to be reliable against transient communication failures. An eventual test against an actually-deployed-to-a-real-GCP-K8S-cluster would be talking over the Internet, but IMO even then I'd like to accept an occasional flake vs trying to add retries to every bit of test-setup code we end up writing -- integration tests against "real" deployed environments tend to be inherently flaky, anyway.

This code also doesn't check that the port-forwarding actually succeeded. If it fails for any reason, this code won't notice. In the worst case, if the forwarding failed because the port to forward was already bound, later code might end up sending requests to the wrong service! I think noticing failure would require writing something like the log-check I suggest--which would give you the wait-until-port-is-forwarded check, too.

Reply:
We're not going to fail to notice that the port-forwarding failed, because the subsequent test will fail because it can't connect to anything to upload reports. I implemented a polling loop that waits until the forwarded port can be connected to and is writable, since that's the observable effect of `kubectl port-forward` we care about. I believe parsing the output of `kubectl` is too brittle to be of much value beyond checking that we can connect to the port.

Reply:
Right: the test failing in this case is bad, because it can happen even if everything is working correctly & the test code happened to run fast enough to get to the first request before the port-forward actually took effect. That's why I want to check that the port-forward completes before continuing.

Reply:
OK, I added a check to the TCP port polling loop that verifies that the child process is still running. Now, if we fail to connect with `ConnectionRefused`, and the child process has exited, we'll immediately bail out of the test. I believe this achieves what you want.